cnd/src/swap_protocols/rfc003/ethereum/htlc_events.rs

```diff
@@ -94,35 +85,20 @@ impl HtlcRefunded<Ethereum, asset::Ether> for Cache<Web3Connector> {
     }
 }
 
-fn calculate_contract_address_from_deployment_transaction(tx: &Transaction) -> Address {
-    tx.from.calculate_contract_address(&tx.nonce)
-}
-
 async fn htlc_redeemed<A: Asset>(
     connector: Cache<Web3Connector>,
     _htlc_params: HtlcParams<Ethereum, A>,
     htlc_deployment: &Deployed<Ethereum>,
     start_of_swap: NaiveDateTime,
 ) -> anyhow::Result<Redeemed<Ethereum>> {
-    let pattern = TransactionPattern {
-        from_address: None,
-        to_address: None,
-        is_contract_creation: None,
-        transaction_data: None,
-        transaction_data_length: None,
-        events: Some(vec![Event {
-            address: Some(htlc_deployment.location),
-            data: None,
-            topics: vec![Some(Topic(*REDEEM_LOG_MSG))],
-        }]),
+    let event = Event {
+        address: Some(htlc_deployment.location),
+        topics: vec![Some(Topic(*REDEEM_LOG_MSG))],
     };
 
-    let TransactionAndReceipt {
-        transaction,
-        receipt,
-    } = matching_transaction(connector, pattern, start_of_swap)
-        .await
-        .context("failed to find transaction to redeem from htlc")?;
+    let (transaction, receipt) =
+        matching_event_transaction(connector, start_of_swap, event, "redeemed").await?;
+
     let log = receipt
```

Comment:
As you can see, we iterate over all the logs and try to find the one again that we passed to `matching_event_transaction`.

We can do a similar optimization as with the contract address and return the event already from the `matching_event_transaction` function. That would save us this, in practice, unreachable error path of trying to find the one log again.